Zend/zend_inheritance.c

```diff
@@ -1898,6 +1898,121 @@ static zend_class_entry *fixup_trait_scope(const zend_function *fn, zend_class_e
 	return fn->common.scope->ce_flags & ZEND_ACC_TRAIT ? ce : fn->common.scope;
 }
 
+/* If the type was resolved then either the zend_string pointer is different, or the zend_type_list is */
+static inline bool zend_was_type_resolved(zend_type original_type, zend_type resolved_type)
+{
+	return original_type.ptr != resolved_type.ptr;
+}
+
+static zend_type zend_resolve_name_type(zend_type type, const zend_class_entry *const ce)
+{
+	ZEND_ASSERT(ZEND_TYPE_HAS_NAME(type));
+	if (ZEND_TYPE_IS_RELATIVE_SELF(type)) {
+		zend_type resolved_type = (zend_type) ZEND_TYPE_INIT_CLASS(zend_string_copy(ce->name), /* allows_null */ false, /* extra_flags */ ZEND_TYPE_FULL_MASK(type));
+		return resolved_type;
+	} else if (ZEND_TYPE_IS_RELATIVE_PARENT(type)) {
+		if (!ce->parent) {
+			zend_error_noreturn(E_COMPILE_ERROR,
+				"Cannot use trait which has \"parent\" as a type when current class scope has no parent");
+			return (zend_type) ZEND_TYPE_INIT_NONE(0);
```

Comment:
I'm actually quite confused by the ``zend_error_noreturn()`` function, does it just... bail? I know if I could just omit the return or I needed one to prevent the execution of the current C function.

Reply:
Yes, `zend_error_noreturn` always bails. You should pass the proper `type` that bails (like `E_COMPILE_ERROR`), and if you fail to do so (which is an error) the function will `abort()`.

Reply:
ACK